### DISCUSSION TITLE
Are the undocumented set operators `comprehension` and `is_singleton` safe to use?

### DISCUSSION THREAD

**Discussion by tnelson** [cvc5/cvc5#11039] [Q&A] [answered]
Sorry to ask a (maybe) silly question, but: we've been working with CVC5's [theory of finite sets and relations](https://cvc5.github.io/docs/cvc5-1.1.2/theories/sets-and-relations.html#). We noticed that there were a number of operators in code and test suite that would be very useful, but do not appear in the [documentation](https://cvc5.github.io/docs/cvc5-1.1.2/theories/sets-and-relations.html#) for the theory of finite sets; `is_singleton` and `comprehension` are especially desirable for us. 

If this is just an oversight in the docs, we'll joyfully start using these operators. Indeed, we've already been experimenting with them in the latest release. But if they are unstable, we might need to find alternative workarounds.  Thanks!

**Comment by mudathirmahgoub** [comment, marked as answer]
We need to update the documentation. But for now, 
operators `set.is_singleton` and `set.is_empty` are simple and stable. 
Operators like `set.filter`, `set.map` are fine to use without cardinality and quantifiers. 
Operators `set.comprehension`, `set.fold` are unstable and their implementation may change. 
Can you use `set.filter` instead of `set.comprehension` in your application?

**Reply by tnelson** [reply]
Much appreciated; this is helpful. I think perhaps we can use `set.filter`, assuming it lets us construct a subset of a set that satisfies a predicate. (For context, the project is in the same vein as some of your own Alloy-to-SMT work, but in our own Alloy-like language "Forge".)

Two follow-ups which I hope are not troublesome:
* Is there a resource we should look at, other than the code and tests, for `set.filter`? The keyword `filter` doesn't appear in the citation `[ BBRT17 ]`, unfortunately.
* One of our target cases requires (at least, naively) combining cardinality and comprehension. Can the problem be avoided if we add a solver variable of appropriate `Set` sort and assert it is equal to the `set.filter`, so there is indirection between `set.card` and `set.filter`, or does the issue arise whenever the two operators appear together in the assertion set?

Thank you again for the reply!

**Reply by mudathirmahgoub** [reply]
- `set.filter` and `bag.filter` operators are documented in this paper https://easychair.org/publications/paper/MWDj 
- This workaround does not help and has the same effect of using `set.filter` only internally. We did not test using `set.card` with `set.filter` in practice, but in theory it can work with sat problems. For unsat problems, we may not terminate since `set.card` may introduce new set elements during model building that were not considered by `set.filter` during solving.
